src/realm/node_header.hpp

```diff
@@ -229,16 +301,631 @@ class NodeHeader {
+    const auto num_elements = word & 0b1111111111;
+    return num_elements;
+}
+
+template <>
+inline size_t NodeHeader::calc_size<NodeHeader::Encoding::Packed>(size_t num_elements, size_t element_size)
+{
+    return NodeHeader::header_size + align_bits_to8(num_elements * element_size);
+}
+template <>
+inline size_t NodeHeader::calc_size<NodeHeader::Encoding::WTypBits>(size_t num_elements, size_t element_size)
+{
+    return calc_byte_size(wtype_Bits, num_elements, static_cast<uint_least8_t>(element_size));
+}
+template <>
+inline size_t NodeHeader::calc_size<NodeHeader::Encoding::WTypMult>(size_t num_elements, size_t element_size)
+{
+    return calc_byte_size(wtype_Multiply, num_elements, static_cast<uint_least8_t>(element_size));
+}
+
+template <>
+inline size_t NodeHeader::calc_size<NodeHeader::Encoding::WTypIgn>(size_t num_elements)
+{
+    return calc_byte_size(wtype_Ignore, num_elements, 0);
+}
+template <>
+inline size_t NodeHeader::calc_size<NodeHeader::Encoding::AofP>(size_t num_elements, size_t elementA_size,
+                                                                size_t elementB_size)
+{
+    return NodeHeader::header_size + align_bits_to8(num_elements * (elementA_size + elementB_size));
+}
+template <>
+inline size_t NodeHeader::calc_size<NodeHeader::Encoding::PofA>(size_t num_elements, size_t elementA_size,
+                                                                size_t elementB_size)
+{
+    return NodeHeader::header_size + align_bits_to8(num_elements * (elementA_size + elementB_size));
+}
+template <>
+inline size_t NodeHeader::calc_size<NodeHeader::Encoding::Flex>(size_t arrayA_num_elements,
+                                                                size_t arrayB_num_elements, size_t elementA_size,
+                                                                size_t elementB_size)
+{
+    return NodeHeader::header_size +
+           align_bits_to8(arrayA_num_elements * elementA_size + arrayB_num_elements * elementB_size);
+}
+
+size_t inline NodeHeader::get_byte_size_from_header(const char* header) noexcept
+{
+    auto h = header;
+
+    auto encoding = get_encoding(h);
+    REALM_ASSERT_DEBUG(encoding >= Encoding::WTypBits && encoding <= Encoding::Flex);
+    switch (encoding) {
+        case Encoding::WTypBits:
+        case Encoding::WTypIgn:
+        case Encoding::WTypMult: {
+            WidthType wtype = get_wtype_from_header(header);
+            size_t width;
+            size_t size;
+            width = get_width_from_header(header);
+            size = get_size_from_header(header);
+            return calc_byte_size(wtype, size, static_cast<uint_least8_t>(width));
+        }
+        case Encoding::Packed:
+            return NodeHeader::header_size + align_bits_to8(get_num_elements<NodeHeader::Encoding::Packed>(h) *
+                                                            get_element_size<NodeHeader::Encoding::Packed>(h));
+        case Encoding::AofP:
```

Comment:
Isn't AofP the same as PofA? We could possibly save some code here?

Reply:
AofP and PofA are not part of this effort, I am going to delete all the logic for them, there is no compressor for those types and in many places we only assert for Flex or Packed.